File: tests/volumes.c
/* Conformance test of the volumes group on Linux: the enumeration is compared with
 * what getmntent reads from the same table, the space of a volume with statvfs,
 * and its format with the type text of the mount that df would name for the path:
 * the longest mount point that leads to it. The providers match devices instead. */
#define _GNU_SOURCE
#include "dotnet_pal.h"
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <limits.h>
#include <mntent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/statvfs.h>
#include <sys/wait.h>
#include <unistd.h>
#ifdef PAL_HOST_TEST
extern int pal_volumes_fault;
#endif
#define INVALID DOTNET_PAL_INVALID_ARGUMENT
#define SMALL DOTNET_PAL_BUFFER_TOO_SMALL
#define MISSING DOTNET_PAL_NOT_FOUND
#define CAPACITY (DOTNET_PAL_MAX_NAME + 1)
#define S(text) (const uint8_t*)(text), strlen(text)
static const dotnet_pal_volumes_ops *v;
static uint8_t out[CAPACITY + 1];
/* What the counters have to say at the end: every call of the group goes through these two. */
static dotnet_pal_volumes_stats expected;
static uint32_t entry(size_t index, uint8_t *buffer, size_t capacity, size_t *needed) {
    uint32_t status = v->entry(index, buffer, capacity, needed);
    ++*(status == 0 ? &expected.entry_ok : &expected.rejected_or_failed); return status;
}
static uint32_t status_of(const char *path, dotnet_pal_volume_status *s) {
    memset(s, 0x55, sizeof *s);
    uint32_t status = v->status(S(path), s, sizeof *s);
    ++*(status == 0 ? &expected.status_ok : &expected.rejected_or_failed); return status;
}
static int zero(const uint8_t *bytes, size_t size) { while (size--) if (*bytes++) return 0; return 1; }
/* The table as the C library reads it. */
struct mount { char *point, *type; };
static struct mount *mounts; static size_t count;
static void read_table(void) {
    for (size_t i = 0; i < count; ++i) { free(mounts[i].point); free(mounts[i].type); }
    count = 0;
    FILE *table = setmntent("/proc/self/mounts", "r"); assert(table);
    for (struct mntent *m; (m = getmntent(table));) {
        mounts = realloc(mounts, (count + 1) * sizeof *mounts); assert(mounts);
        mounts[count].point = strdup(m->mnt_dir); mounts[count].type = strdup(m->mnt_type);
        assert(mounts[count].point && mounts[count].type); ++count;
    }
    endmntent(table);
}
/* The type text of the longest mount point that leads to a resolved path, the last of them where one covers another. */
static const char *holder(const char *path) {
    const char *type = NULL; size_t longest = 0;
    for (size_t i = 0; i < count; ++i) {
        size_t length = strlen(mounts[i].point);
        int leads = strncmp(path, mounts[i].point, length) == 0 && (length == 1 || path[length] == 0 || path[length] == '/');
        if (leads && length >= longest) { longest = length; type = mounts[i].type; }
    }
    assert(type); return type;
}
/* Entry `index` is `text`: whole with its NUL where it fits, cleared behind it and never past the capacity; the length alone, and a cleared buffer, where it does not. */
static void delivers(size_t index, const char *text) {
    size_t length = strlen(text) + 1, needed = 7;
    assert(length >= 2 && length <= CAPACITY);
    memset(out, 0xAA, sizeof out);
    assert(entry(index, out, CAPACITY, &needed) == 0 && needed == length && memcmp(out, text, length) == 0 && zero(out + length, CAPACITY - length) && out[CAPACITY] == 0xAA);
    memset(out, 0xAA, sizeof out); needed = 7;
    assert(entry(index, out, length, &needed) == 0 && needed == length && memcmp(out, text, length) == 0 && out[length] == 0xAA);
    memset(out, 0xAA, sizeof out); needed = 7;
    assert(entry(index, out, length - 1, &needed) == SMALL && needed == length && zero(out, length - 1) && out[length - 1] == 0xAA);
    needed = 7;
    assert(entry(index, NULL, 0, &needed) == SMALL && needed == length);
}
static void refuses(size_t index, uint32_t status) {
    size_t needed = 7;
    memset(out, 0xAA, sizeof out);
    assert(entry(index, out, CAPACITY, &needed) == status && needed == 0 && zero(out, CAPACITY) && out[CAPACITY] == 0xAA);
}
/* The status of a path is what statvfs measures and the type text of the mount that holds it. Free space moves under a running
 * system, so the numbers are compared once two measurements around the call agree. Returns the capacity. */
static uint64_t measured(const char *path, const char *type) {
    dotnet_pal_volume_status s; struct statvfs earlier, later; char resolved[PATH_MAX];
    for (int attempt = 0;; ++attempt) {
        assert(statvfs(path, &earlier) == 0 && status_of(path, &s) == 0 && statvfs(path, &later) == 0);
        if (earlier.f_blocks == later.f_blocks && earlier.f_bfree == later.f_bfree && earlier.f_bavail == later.f_bavail) break;
        assert(attempt < 50);
    }
    uint64_t unit = earlier.f_frsize;
    assert(s.total_bytes == earlier.f_blocks * unit && s.free_bytes == earlier.f_bfree * unit && s.available_bytes == earlier.f_bavail * unit);
    assert(s.available_bytes <= s.free_bytes && s.free_bytes <= s.total_bytes);
    if (!type) { assert(realpath(path, resolved)); type = holder(resolved); }
    size_t length = strlen(type);
    assert(length > 0 && length < sizeof s.format && memcmp(s.format, type, length) == 0 && zero(s.format + length, sizeof s.format - length));
    return s.total_bytes;
}
static void missing(const char *path, int code, uint32_t status) {
    dotnet_pal_volume_status s; struct statvfs k;
    errno = 0;
    assert(statvfs(path, &k) == -1 && errno == code && status_of(path, &s) == status && zero((const uint8_t*)&s, sizeof s));
}
/* ACCESS_DENIED needs a caller the permission bits apply to: a child that gives its privileges up. 77 when the kernel lets it through. */
static int denied(const char *path) {
    struct statvfs k;
    if (setgroups(0, NULL) != 0 || setgid(54321) != 0 || setuid(54321) != 0) return 77;
    if (statvfs(path, &k) == 0 || errno != EACCES) return 77;
    missing(path, EACCES, DOTNET_PAL_ACCESS_DENIED);
    return 0;
}
int main(int argc, char **argv) {
#ifdef PAL_HOST_TEST
    pal_volumes_fault = argc > 1 ? atoi(argv[1]) : 0;
#else
    (void)argc; (void)argv;
#endif
    const dotnet_pal_api *api = dotnet_pal_get_api(2);
#ifdef PAL_HOST_TEST
    if (pal_volumes_fault == 1) { assert(!api); puts("VOLUMES malformed host rejected"); return 0; }
#endif
    assert(api && api->header.struct_size >= DOTNET_PAL_VOLUMES_API_SIZE);
    assert(api->header.capabilities & DOTNET_PAL_CAP_VOLUMES);
    v = &api->volumes;
    assert(v->entry && v->status && v->read_stats);
    dotnet_pal_volumes_stats before, after; dotnet_pal_volume_status s; size_t needed = 7;
#ifdef PAL_HOST_TEST
    if (pal_volumes_fault == 2) {
        /* A text without its NUL, with one inside, empty, without a length, behind a status the call does not have, longer than any path or
         * "too small" for a buffer it fits is refused and cleared; the end is the enumeration's to report, without the outputs of that call. */
        for (int i = 0; i < 8; ++i) { needed = 7; memset(out, 0xAA, sizeof out); assert(v->entry(0, out, CAPACITY, &needed) == DOTNET_PAL_OS_ERROR && needed == 0 && zero(out, CAPACITY)); }
        needed = 7; memset(out, 0xAA, sizeof out);
        assert(v->entry(0, out, CAPACITY, &needed) == MISSING && needed == 0 && zero(out, CAPACITY));
        /* More free than there is, more usable than free, a format without its end and statuses the call does not have are no answer. */
        for (int i = 0; i < 5; ++i) { memset(&s, 0x55, sizeof s); assert(v->status(S("/"), &s, sizeof s) == DOTNET_PAL_OS_ERROR && zero((const uint8_t*)&s, sizeof s)); }
        memset(&s, 0x55, sizeof s);
        assert(v->status(S("/"), &s, sizeof s) == MISSING && zero((const uint8_t*)&s, sizeof s));
        assert(v->read_stats(&after, sizeof after) == 0 && after.rejected_or_failed == 15 && after.entry_ok + after.status_ok == 0);
        puts("VOLUMES host errors sanitized"); return 0;
    }
#endif
    char root[] = "/tmp/pal-volumes-XXXXXX", path[PATH_MAX];
    assert(mkdtemp(root));
    /* A volume of this test's own where the process may mount: a megabyte of tmpfs on a path with every byte the table escapes. */
    snprintf(path, sizeof path, "%s/a b\tc\\d\ne", root);
    assert(mkdir(path, 0700) == 0);
    int escaped = mount("pal-volumes", path, "tmpfs", 0, "size=1m") == 0;
    assert(escaped || errno == EPERM);

    /* Enumeration: the mount points of the table in its order, NOT_FOUND from the count on. Each is a path that is there and on a volume. */
    read_table();
    int has_root = 0, own = 0;
    for (size_t i = 0; i < count; ++i) {
        struct stat node; struct statvfs space;
        delivers(i, mounts[i].point);
        assert(mounts[i].point[0] == '/' && lstat(mounts[i].point, &node) == 0 && statvfs(mounts[i].point, &space) == 0);
        if (strcmp(mounts[i].point, "/") == 0) has_root = 1;
        if (strcmp(mounts[i].point, path) == 0) ++own;
    }
    assert(count > 0 && has_root && own == escaped);
    refuses(count, MISSING); refuses(count + 1, MISSING); refuses(SIZE_MAX, MISSING);

    /* status: the root, the scratch directory, procfs, a file, and every mount point, among them those that are files and those another covers. */
    uint64_t total = measured("/", NULL);
    measured(root, NULL); measured("/proc", "proc"); measured("/proc/self/status", "proc"); measured("/etc/passwd", NULL); measured(".", NULL);
    for (size_t i = 0; i < count; ++i) measured(mounts[i].point, NULL);
    if (escaped) assert(measured(path, "tmpfs") == 1024 * 1024);
    /* A path that names nothing, directly or through a file. */
    missing("/no/such/pal/volume", ENOENT, MISSING); missing("/etc/passwd/x", ENOTDIR, MISSING);
    int refused = -1;
    if (geteuid() == 0) {
        snprintf(path, sizeof path, "%s/locked", root);
        assert(mkdir(path, 0700) == 0);
        strcat(path, "/inside");
        pid_t child = fork(); assert(child >= 0);
        if (child == 0) _exit(denied(path));
        int result = 0;
        assert(waitpid(child, &result, 0) == child && WIFEXITED(result) && (WEXITSTATUS(result) == 0 || WEXITSTATUS(result) == 77));
        refused = WEXITSTATUS(result) == 0;
    }

    /* Argument validation: refused before a provider runs, with one count each and nothing else. */
    assert(v->read_stats(&before, sizeof before) == 0);
    assert(v->entry(0, out, CAPACITY, NULL) == INVALID && v->entry(0, out, CAPACITY, (size_t*)((char*)&needed + 1)) == INVALID && v->entry(0, NULL, 8, &needed) == INVALID);
    assert(v->entry(0, out, (size_t)PTRDIFF_MAX + 1, &needed) == INVALID && v->entry(0, (uint8_t*)(UINTPTR_MAX - 3), 8, &needed) == INVALID && needed == 7);
    memset(&s, 0x55, sizeof s);
    assert(v->status(S("/"), NULL, sizeof s) == INVALID && v->status(S("/"), (void*)((char*)&s + 1), sizeof s) == INVALID && v->status(S("/"), &s, sizeof s - 1) == INVALID && s.total_bytes == UINT64_C(0x5555555555555555));
    /* A path is 1 to MAX_NAME bytes without a NUL; the output is cleared once it is known to be one. */
    static uint8_t longest[DOTNET_PAL_MAX_NAME + 1];
    memset(longest, 'a', sizeof longest);
    const struct { const uint8_t *path; size_t length; } paths[] = {{NULL, 1}, {(const uint8_t*)"/", 0}, {longest, sizeof longest}, {(const uint8_t*)"/a\0b", 4}, {(const uint8_t*)(UINTPTR_MAX - 1), 8}};
    for (size_t i = 0; i < 5; ++i) { memset(&s, 0x55, sizeof s); assert(v->status(paths[i].path, paths[i].length, &s, sizeof s) == INVALID && zero((const uint8_t*)&s, sizeof s)); }
    /* The longest path the boundary takes reaches the provider, where the kernel has the word. */
    memset(&s, 0x55, sizeof s);
    assert(v->status(longest, DOTNET_PAL_MAX_NAME, &s, sizeof s) != INVALID && zero((const uint8_t*)&s, sizeof s));
    assert(v->read_stats(NULL, sizeof after) == INVALID && v->read_stats(&after, sizeof after - 1) == INVALID);
    assert(v->read_stats(&after, sizeof after) == 0 && after.rejected_or_failed == before.rejected_or_failed + 14 && after.entry_ok == before.entry_ok && after.status_ok == before.status_ok);
    expected.rejected_or_failed += 14;

    /* Counters: one per answered question of each kind, one for everything refused. */
    assert(entry(0, out, CAPACITY, &needed) == 0 && status_of("/", &s) == 0 && entry(count, out, CAPACITY, &needed) == MISSING && entry(0, out, 1, &needed) == SMALL);
    assert(status_of("/no/such/pal/volume", &s) == MISSING);
    assert(v->read_stats(&after, sizeof after) == 0);
    assert(after.entry_ok == before.entry_ok + 1 && after.status_ok == before.status_ok + 1 && after.rejected_or_failed == before.rejected_or_failed + 17);
    assert(after.entry_ok == expected.entry_ok && after.status_ok == expected.status_ok && after.rejected_or_failed == expected.rejected_or_failed);
    assert(after.entry_ok == 2 * count + 1 && after.rejected_or_failed == 2 * count + 3 + 2 + 17);

    if (escaped) { snprintf(path, sizeof path, "%s/a b\tc\\d\ne", root); assert(umount(path) == 0); }
    snprintf(path, sizeof path, "%s/a b\tc\\d\ne", root); assert(rmdir(path) == 0);
    if (geteuid() == 0) { snprintf(path, sizeof path, "%s/locked", root); assert(rmdir(path) == 0); }
    assert(rmdir(root) == 0);
    printf("VOLUMES PASS mounts=%zu root=%s scratch=%s total_mib=%llu statuses=%llu refused=%llu escaped_mount=%s access_denied=%s\n", count, holder("/"), holder(root),
        (unsigned long long)(total >> 20), (unsigned long long)after.status_ok, (unsigned long long)after.rejected_or_failed, escaped ? "checked" : "not permitted", refused == 1 ? "checked" : "not checked");
    return 0;
}

File: crates/dotnet-pal-std/tests/volumes.rs
//! Exercises the volumes group of the std port through the negotiated C table, on
//! whatever desktop OS runs the test, against what the C library lists and
//! measures itself: getmntent and statvfs on Linux, getmntinfo and statfs on macOS.
use dotnet_pal_rs::runtime::MAX_NAME;
#[cfg(any(target_os = "linux", target_os = "macos"))]
use dotnet_pal_rs::runtime::{BUFFER_TOO_SMALL, NOT_FOUND};
use dotnet_pal_rs::volumes::{self, Status};
use dotnet_pal_rs::{INVALID_ARGUMENT, OK};
use std::{mem::size_of, ptr, sync::{Mutex, MutexGuard}};

const CAPACITY: usize = MAX_NAME + 1;
/// One test at a time: the counters are the process's.
static SERIAL: Mutex<()> = Mutex::new(());
fn serial() -> MutexGuard<'static, ()> { SERIAL.lock().unwrap_or_else(|poisoned| poisoned.into_inner()) }
fn ops() -> &'static volumes::Ops {
    let api = dotnet_pal_std::api();
    assert!(!api.is_null());
    let api = unsafe { &*api };
    assert_eq!(api.header.capabilities & volumes::CAP, volumes::CAP);
    &api.volumes
}
/// Status, needed length and the whole buffer of one call; no buffer at all for capacity 0.
fn entry(index: usize, capacity: usize) -> (u32, usize, Vec<u8>) {
    let (mut needed, mut buffer) = (usize::MAX, vec![0xAAu8; capacity]);
    let status = unsafe { ops().entry.unwrap()(index, if capacity == 0 { ptr::null_mut() } else { buffer.as_mut_ptr() }, capacity, &mut needed) };
    (status, needed, buffer)
}
/// Status of a call and what it left in an output that started as garbage.
fn status(path: &[u8]) -> (u32, Status) {
    let mut out = Status { total_bytes: 7, free_bytes: 7, available_bytes: 7, format: [0xAA; 32] };
    (unsafe { ops().status.unwrap()(path.as_ptr(), path.len(), &mut out, size_of::<Status>()) }, out)
}
fn cleared(s: &Status) -> bool { s.total_bytes == 0 && s.free_bytes == 0 && s.available_bytes == 0 && s.format == [0; 32] }
fn stats() -> volumes::Stats {
    let mut out = volumes::Stats::default();
    assert_eq!(unsafe { ops().read_stats.unwrap()(&mut out, size_of::<volumes::Stats>()) }, OK);
    out
}
fn counted(s: &volumes::Stats) -> (u64, u64, u64) { (s.entry_ok, s.status_ok, s.rejected_or_failed) }

/// The mount points as the C library lists them, with the type text of each.
#[cfg(target_os = "linux")]
fn listed() -> Vec<(Vec<u8>, Vec<u8>)> {
    let table = unsafe { libc::setmntent(c"/proc/self/mounts".as_ptr(), c"r".as_ptr()) };
    assert!(!table.is_null());
    let mut all = Vec::new();
    loop {
        let mount = unsafe { libc::getmntent(table) };
        if mount.is_null() { break; }
        all.push(unsafe { (std::ffi::CStr::from_ptr((*mount).mnt_dir).to_bytes().to_vec(), std::ffi::CStr::from_ptr((*mount).mnt_type).to_bytes().to_vec()) });
    }
    unsafe { libc::endmntent(table) };
    all
}
#[cfg(target_os = "macos")]
fn listed() -> Vec<(Vec<u8>, Vec<u8>)> {
    let text = |field: &[std::ffi::c_char]| field.iter().map(|c| *c as u8).take_while(|byte| *byte != 0).collect::<Vec<u8>>();
    let mut list = ptr::null_mut();
    let count = unsafe { libc::getmntinfo(&mut list, libc::MNT_NOWAIT) };
    assert!(count > 0);
    unsafe { std::slice::from_raw_parts(list, count as usize) }.iter().map(|mount| (text(&mount.f_mntonname), text(&mount.f_fstypename))).collect()
}
/// Capacity, free and available bytes and, where the call names it, the format, as the C library measures them for a path.
#[cfg(target_os = "linux")]
fn measured(path: &[u8]) -> (u64, u64, u64, Option<Vec<u8>>) {
    let (path, mut space) = (std::ffi::CString::new(path).unwrap(), unsafe { std::mem::zeroed::<libc::statvfs>() });
    assert_eq!(unsafe { libc::statvfs(path.as_ptr(), &mut space) }, 0);
    let unit = space.f_frsize as u64;
    (space.f_blocks as u64 * unit, space.f_bfree as u64 * unit, space.f_bavail as u64 * unit, None)
}
#[cfg(target_os = "macos")]
fn measured(path: &[u8]) -> (u64, u64, u64, Option<Vec<u8>>) {
    let (path, mut space) = (std::ffi::CString::new(path).unwrap(), unsafe { std::mem::zeroed::<libc::statfs>() });
    assert_eq!(unsafe { libc::statfs(path.as_ptr(), &mut space) }, 0);
    let unit = u64::from(space.f_bsize);
    (space.f_blocks * unit, space.f_bfree * unit, space.f_bavail * unit, Some(space.f_fstypename.iter().map(|c| *c as u8).take_while(|byte| *byte != 0).collect()))
}
/// The type text of the mount that holds a resolved path, found the way df finds it: the longest mount point that leads to the
/// path, the last of them where one covers another. The provider matches devices instead, so the two are independent.
#[cfg(target_os = "linux")]
fn holder(path: &[u8]) -> Vec<u8> {
    let leads = |point: &[u8]| path.starts_with(point) && (point == b"/" || path.len() == point.len() || path[point.len()] == b'/');
    listed().into_iter().filter(|(point, _)| leads(point)).max_by_key(|(point, _)| point.len()).map(|(_, kind)| kind).unwrap()
}

#[cfg(any(target_os = "linux", target_os = "macos"))]
#[test]
fn mount_points_enumerate_as_the_c_library_lists_them() {
    let _serial = serial();
    // A volume may come or go while the test runs. Only a table that stood still, the same before and after, is compared.
    for attempt in 0.. {
        let (before, earlier) = (stats(), listed());
        let mut points = Vec::new();
        loop {
            let (code, needed, buffer) = entry(points.len(), CAPACITY);
            if code == NOT_FOUND { assert_eq!((needed, buffer), (0, vec![0; CAPACITY])); break; }
            assert_eq!(code, OK);
            // The text, its NUL, and a cleared rest.
            assert!(needed >= 2 && buffer[needed - 1] == 0 && !buffer[..needed - 1].contains(&0) && buffer[needed..].iter().all(|byte| *byte == 0));
            points.push(buffer[..needed - 1].to_vec());
        }
        let later = listed();
        if earlier != later { assert!(attempt < 20, "the mount table never stood still"); continue; }
        assert_eq!(points, earlier.iter().map(|(point, _)| point.clone()).collect::<Vec<_>>());
        assert!(points.iter().any(|point| point == b"/"));
        for point in &points {
            use std::os::unix::ffi::OsStrExt;
            assert_eq!(point[0], b'/');
            assert!(std::fs::symlink_metadata(std::ffi::OsStr::from_bytes(point)).is_ok(), "{}", String::from_utf8_lossy(point));
            measured(point);
        }
        // Each delivers whole where it fits and its length alone, with a cleared buffer, where it does not.
        for (index, point) in points.iter().enumerate() {
            let needed = point.len() + 1;
            let mut whole = point.clone();
            whole.push(0);
            assert_eq!(entry(index, needed), (OK, needed, whole));
            assert_eq!(entry(index, needed - 1), (BUFFER_TOO_SMALL, needed, vec![0; needed - 1]));
            assert_eq!(entry(index, 0), (BUFFER_TOO_SMALL, needed, Vec::new()));
        }
        for index in [points.len() + 1, usize::MAX] { assert_eq!(entry(index, CAPACITY), (NOT_FOUND, 0, vec![0; CAPACITY])); }
        if listed() != later { assert!(attempt < 20, "the mount table never stood still"); continue; }
        let count = points.len() as u64;
        assert_eq!(counted(&stats()), (before.entry_ok + 2 * count, before.status_ok, before.rejected_or_failed + 2 * count + 3));
        break;
    }
}

#[cfg(any(target_os = "linux", target_os = "macos"))]
#[test]
fn status_is_the_space_and_the_format_of_the_volume_that_holds_a_path() {
    use std::os::unix::ffi::OsStrExt;
    let _serial = serial();
    let before = stats();
    let scratch = std::env::temp_dir().canonicalize().unwrap();
    let paths: Vec<Vec<u8>> = vec![b"/".to_vec(), scratch.as_os_str().as_bytes().to_vec(), if cfg!(target_os = "linux") { b"/proc".to_vec() } else { b"/dev".to_vec() }];
    let mut asked = 0;
    for path in &paths {
        // Free space moves under a running system: the numbers are compared while two measurements around the call agree.
        for attempt in 0.. {
            let (earlier, (code, s), later) = (measured(path), status(path), measured(path));
            asked += 1;
            assert_eq!(code, OK);
            if earlier != later { assert!(attempt < 50, "the free space of {} never stood still", String::from_utf8_lossy(path)); continue; }
            assert_eq!((s.total_bytes, s.free_bytes, s.available_bytes), (earlier.0, earlier.1, earlier.2), "{}", String::from_utf8_lossy(path));
            assert!(s.available_bytes <= s.free_bytes && s.free_bytes <= s.total_bytes);
            let length = s.format.iter().position(|byte| *byte == 0).unwrap();
            assert!(s.format[length..].iter().all(|byte| *byte == 0));
            #[cfg(target_os = "linux")]
            assert_eq!(s.format[..length], holder(path)[..], "{}", String::from_utf8_lossy(path));
            #[cfg(target_os = "macos")]
            assert_eq!(Some(s.format[..length].to_vec()), earlier.3, "{}", String::from_utf8_lossy(path));
            assert!(length > 0);
            break;
        }
    }
    if cfg!(target_os = "linux") { assert_eq!(status(b"/proc").1.format[..5], *b"proc\0"); asked += 1; }
    // A path that names nothing, directly or through a file, leaves a cleared status.
    let file = scratch.join(format!("pal-std-volumes-{}", std::process::id()));
    std::fs::write(&file, b"x").unwrap();
    let through = [file.as_os_str().as_bytes(), b"/x"].concat();
    for missing in [&b"/no/such/pal/volume"[..], &through] {
        let (code, s) = status(missing);
        assert!(code == NOT_FOUND && cleared(&s));
    }
    // A file is on a volume as much as a directory is.
    let (code, s) = status(file.as_os_str().as_bytes());
    assert!(code == OK && s.total_bytes == measured(file.as_os_str().as_bytes()).0);
    std::fs::remove_file(&file).unwrap();
    assert_eq!(counted(&stats()), (before.entry_ok, before.status_ok + asked + 1, before.rejected_or_failed + 2));
}

#[test]
fn malformed_arguments_are_refused_before_the_provider() {
    let _serial = serial();
    let (v, before) = (ops(), stats());
    let (mut needed, mut buffer) = (7usize, [0xAAu8; 8]);
    assert_eq!(unsafe { v.entry.unwrap()(0, buffer.as_mut_ptr(), 8, ptr::null_mut()) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.entry.unwrap()(0, buffer.as_mut_ptr(), 8, (&raw mut needed).cast::<u8>().wrapping_add(1).cast()) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.entry.unwrap()(0, ptr::null_mut(), 8, &mut needed) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.entry.unwrap()(0, buffer.as_mut_ptr(), isize::MAX as usize + 1, &mut needed) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.entry.unwrap()(0, (usize::MAX - 3) as *mut u8, 8, &mut needed) }, INVALID_ARGUMENT);
    assert_eq!((needed, buffer), (7, [0xAA; 8]));
    let mut out = Status { total_bytes: 7, free_bytes: 7, available_bytes: 7, format: [0xAA; 32] };
    assert_eq!(unsafe { v.status.unwrap()(b"/".as_ptr(), 1, ptr::null_mut(), size_of::<Status>()) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.status.unwrap()(b"/".as_ptr(), 1, (&raw mut out).cast::<u8>().wrapping_add(1).cast(), size_of::<Status>()) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.status.unwrap()(b"/".as_ptr(), 1, &mut out, size_of::<Status>() - 1) }, INVALID_ARGUMENT);
    assert_eq!(out.total_bytes, 7);
    // A path is 1 to MAX_NAME bytes without a NUL; the output is cleared once it is known to be one.
    let long = vec![b'a'; MAX_NAME + 1];
    for (path, length) in [(ptr::null(), 1), (b"/".as_ptr(), 0), (long.as_ptr(), long.len()), (b"/a\0b".as_ptr(), 4), ((usize::MAX - 1) as *const u8, 8)] {
        out = Status { total_bytes: 7, free_bytes: 7, available_bytes: 7, format: [0xAA; 32] };
        assert_eq!(unsafe { v.status.unwrap()(path, length, &mut out, size_of::<Status>()) }, INVALID_ARGUMENT);
        assert!(cleared(&out));
    }
    assert_eq!(unsafe { v.read_stats.unwrap()(ptr::null_mut(), size_of::<volumes::Stats>()) }, INVALID_ARGUMENT);
    assert_eq!(unsafe { v.read_stats.unwrap()(&mut volumes::Stats::default(), size_of::<volumes::Stats>() - 1) }, INVALID_ARGUMENT);
    // Targets without a provider behind the table: every well-formed question is UNSUPPORTED and leaves nothing.
    #[cfg(not(any(target_os = "linux", target_os = "android", target_vendor = "apple")))]
    {
        assert_eq!(entry(0, CAPACITY), (dotnet_pal_rs::UNSUPPORTED, 0, vec![0; CAPACITY]));
        let (code, s) = status(b"/");
        assert!(code == dotnet_pal_rs::UNSUPPORTED && cleared(&s));
    }
    let unsupported = if cfg!(any(target_os = "linux", target_os = "android", target_vendor = "apple")) { 0 } else { 2 };
    assert_eq!(counted(&stats()), (before.entry_ok, before.status_ok, before.rejected_or_failed + 13 + unsupported));
}
